ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/Round.java

```diff
@@ -0,0 +1,139 @@
+    return val == null ? null : Math.round(val);
+  }
+
+  @Udf
+  public Double round(@UdfParameter final Double val, @UdfParameter final Integer decimalPlaces) {
+    // Note that we MUST BigDecimal.valueOf(double) to create the BD to ensure correctness
+    // as this methods initialises via the string representation
+    return val == null
+        ?
+        null : roundBigDecimal(BigDecimal.valueOf(val), decimalPlaces).doubleValue();
+  }
+
+  @Udf(schemaProvider = "provideDecimalSchema")
+  public BigDecimal round(@UdfParameter final BigDecimal val) {
+    return val == null ? null : round(val, 0);
+  }
+
+  @Udf(schemaProvider = "provideDecimalSchemaWithDecimalPlaces")
+  public BigDecimal round(
+      @UdfParameter final BigDecimal val,
+      @UdfParameter
+      final Integer decimalPlaces
+  ) {
+    return val == null ? null : roundBigDecimal(val, decimalPlaces);
+  }
+
+  @UdfSchemaProvider
+  public SqlType provideDecimalSchemaWithDecimalPlaces(final List<SqlType> params) {
+    if (params.size() != 2) {
+      throw new KsqlException("Round udf with decimal places accepts two parameters");
+    }
+    final SqlType s0 = params.get(0);
+    if (s0.baseType() != SqlBaseType.DECIMAL) {
+      throw new KsqlException("The schema provider method for round expects a BigDecimal parameter"
+          + "type as first parameter.");
+    }
+    final SqlType s1 = params.get(1);
+    if (s1.baseType() != SqlBaseType.INTEGER) {
+      throw new KsqlException("The schema provider method for round expects an Integer parameter"
+          + "type as second parameter.");
+    }
+    return s0;
+  }
+
+  @UdfSchemaProvider
+  public SqlType provideDecimalSchema(final List<SqlType> params) {
+    if (params.size() != 1) {
+      throw new KsqlException("Round udf accepts one parameter");
+    }
+    final SqlType s0 = params.get(0);
+    if (s0.baseType() != SqlBaseType.DECIMAL) {
+      throw new KsqlException("The schema provider method for round expects a BigDecimal parameter"
+          + "type as a parameter.");
+    }
+    return s0;
+  }
+
+  /*
+  Unfortunately there is an inconsistency in the way that java.lang.Math and
+  BigDecimal work with respect to rounding:
+
+  Math.round(1.5) --> 2
+  Math.round(-1.5) -- -1
+
+  But:
+
+  new BigDecimal("1.5").setScale(2, RoundingMode.HALF_UP) --> 2
+  new BigDecimal("-1.5").setScale(2, RoundingMode.HALF_UP) --> -2
+
+  There isn't any BigDecimal rounding mode which captures the java.lang.Math behaviour so
+  we need to use different rounding modes on BigDecimal depending on whether the value
+  is +ve or -ve to get consistent behaviour.
+
+  The alternative would be to use the BigDecimal rounding behaviour everywhere but that would
+  mean worse performance in the areas where we currently use java.lang.Math.round().
+  */
+  private BigDecimal roundBigDecimal(final BigDecimal val, final int decimalPlaces) {
+    return val.setScale(decimalPlaces,
```

Comment:
nit: style - would you mind pulling the rounding mode out into a var to make the code more readable, please?

```java
final RoundingMode mode = val.compareTo(BigDecimal.ZERO) > 0 
     ? RoundingMode.HALF_UP 
     : RoundingMode.HALF_DOWN);

return val.setScale(decimalPlaces, roundingMode);
```

Reply:
Also, consider putting this inline...

Reply:
Imho, I don't think it adds much to split, but will change.

Reply:
done